dikastes.go

```diff
@@ -1,44 +1,163 @@
+	"github.com/projectcalico/app-policy/policystore"
+	"github.com/projectcalico/app-policy/syncher"
 	log "github.com/sirupsen/logrus"
 	"google.golang.org/grpc"
-	"google.golang.org/grpc/reflection"
 )
 
 const usage = `Dikastes - the decider.
 
 Usage:
   dikastes server [options]
+  dikastes client <namespace> <account> [--method <method>] [options]
 
 Options:
-  -h --help    Show this screen.`
+  <namespace>            Service account namespace.
+  <account>              Service account name.
+  -h --help              Show this screen.
+  -l --listen <port>     Unix domain socket path [default: /var/run/dikastes/dikastes.sock]
+  -d --dial <target>     Target to dial. [default: localhost:50051]
+  --debug                Log at Debug level.`
 const version = "0.1"
-const default_port = ":50051"
-
+const NODE_NAME_ENV = "K8S_NODENAME"
 
 func main() {
 	arguments, err := docopt.Parse(usage, nil, true, version, false)
 	if err != nil {
 		println(usage)
 		return
 	}
+	if arguments["--debug"].(bool) {
+		log.SetLevel(log.DebugLevel)
+	}
 	if arguments["server"].(bool) {
-		lis, err := net.Listen("tcp", default_port)
+		runServer(arguments)
+	} else if arguments["client"].(bool) {
+		runClient(arguments)
+	}
+	return
+}
+
+func runServer(arguments map[string]interface{}) {
+	filePath := arguments["--listen"].(string)
+	_, err := os.Stat(filePath)
+	if !os.IsNotExist(err) {
+		// file exists, try to delete it.
+		err := os.Remove(filePath)
 		if err != nil {
-			log.Fatalf("Unable to listen on %v")
+			log.WithFields(log.Fields{
+				"listen": filePath,
+				"err":    err,
+			}).Fatal("File exists and unable to remove.")
 		}
-		s := grpc.NewServer()
-		authz.RegisterAuthorizationServer(s, server.NewServer())
-		reflection.Register(s)
-		if err := s.Serve(lis); err != nil {
+	}
+	lis, err := net.Listen("unix", filePath)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"listen": filePath,
+			"err":    err,
+		}).Fatal("Unable to listen.")
+	}
+	defer lis.Close()
+	err = os.Chmod(filePath, 0777) // Anyone on system can connect.
+	if err != nil {
+		log.Fatal("Unable to set write permission on socket.")
+	}
+	gs := grpc.NewServer()
+	store := policystore.NewPolicyStore()
+	checkServer := checker.NewServer(store)
+	authz.RegisterAuthorizationServer(gs, checkServer)
+
+	// Synchronize the policy store
+	opts := getDialOptions()
+	syncClient := syncher.NewClient(arguments["--dial"].(string), opts)
+	syncContext, cancelSync := context.WithCancel(context.Background())
+	defer cancelSync()
+	go syncClient.Sync(syncContext, store)
+
+	// Run gRPC server on separate goroutine so we catch any signals and clean up.
+	go func() {
+		if err := gs.Serve(lis); err != nil {
 			log.Fatalf("failed to serve: %v", err)
 		}
+	}()
+
+	// Use a buffered channel so we don't miss any signals
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, os.Kill, syscall.SIGTERM)
+
+	// Block until a signal is received.
+	s := <-c
+	log.Infof("Got signal:", s)
```

Comment:
I think you can just write `"Got signal:", <-c)` and don't need the `s:= <-c` since it will block until there's something in the channel

Reply:
+1